### PR TITLE
Gun Eyecandy | Muzzle Flash

### PR DESCRIPTION
**Pull request by KitsuneQueen**
## About The Pull Request
Adds muzzle flash to ballistic and energy guns, with a colour for each ammo type. 

## Why It's Good For The Game
gun-shooty

More eyecandy! 

I have a feeling this **will** impact gameplay, however. I've tried to step on as little toes as possible, so stealthy weapons do NOT have a muzzle flash. Suppressed weapons, all of them, will NOT have a muzzle flash, and are completely light-less when firing. The ebow and riot foam toy weapons also have no muzzle flash. 

This does mean firing a revolver or similar dangerous weapons in the dark will give away your position, as before you'd still be hidden. 

Technically this is a balance change, I guess?

---

# Videos!

Regular guns:
https://github.com/tgstation/tgstation/assets/70779744/8a6abc32-7cf8-4120-9027-fa46f22610ec

Suppressor:
https://github.com/tgstation/tgstation/assets/70779744/a19df328-d953-4a7e-a4ff-447825f4a677

---

I'm still incredibly new to coding, so this will probably need way more edits, plus the light values are still up in the air. The flash used to be way bigger and brighter like real-life, but that really sucked, so it's now more subtle.

## Changelog

:cl:
add: Muzzle flash to most guns.
add: Suppressor removes muzzle flash from guns.
/:cl:

### CODE REVIEW

**Comment by WinterDarkraven** [comment]
Oh no the commits.

**Comment by Bibarra** [comment]
How does the flashing look with rapid fire weapons, particularly with Two Time Root Beer's fire rate increase active? Would that flashing be too much for those with photosensitivity?

**Comment by WinterDarkraven** [comment]
> How does the flashing look with rapid fire weapons, particularly with Two Time Root Beer's fire rate increase active? Would that flashing be too much for those with photosensitivity?

Unsure. Videos will be below. There probably should be a photosensitivity setting which should disable client side muzzle flash and stuff.

I don't know how to do client side stuff yet, but it's a very good point!

This is how it looks from a simple test:
https://github.com/tgstation/tgstation/assets/70779744/7cc08311-d39f-48eb-961f-e4b92a9916ef

**Comment by Singul0** [comment]
SOVL, thinking of. does ballistics also already do this? if so why not add this to them too?

**Comment by WinterDarkraven** [comment]
> SOVL, thinking of. does ballistics also already do this? if so why not add this to them too?

Ballistic weapons did not and never have had a muzzle flash. This adds muzzle flash to ballistic and energy weapons.

**Comment by Admiral-Gigaton** [comment]
Make sure this doesn't affect
https://github.com/tgstation/tgstation/pull/78128#issue-1879268689

**Comment by WinterDarkraven** [comment]
> Make sure this doesn't affect [#78128 (comment)](https://github.com/tgstation/tgstation/pull/78128#issue-1879268689)

Thanks for the heads up. Will be removing muzzle flash from it!

**Comment by Admiral-Gigaton** [comment]
> > Make sure this doesn't affect [#78128 (comment)](https://github.com/tgstation/tgstation/pull/78128#issue-1879268689)
> 
> Thanks for the heads up. Will be removing muzzle flash from it!

do turrets get the flash?

**Comment by WinterDarkraven** [comment]
> > > Make sure this doesn't affect [#78128 (comment)](https://github.com/tgstation/tgstation/pull/78128#issue-1879268689)
> > 
> > 
> > Thanks for the heads up. Will be removing muzzle flash from it!
> 
> do turrets get the flash?

AI, nukie, and other turrets do not have a muzzle flash, sadly. That might be a bit more complex to add than standard guns.

**Comment by Admiral-Gigaton** [comment]
> > > > Make sure this doesn't affect [#78128 (comment)](https://github.com/tgstation/tgstation/pull/78128#issue-1879268689)
> > > 
> > > 
> > > Thanks for the heads up. Will be removing muzzle flash from it!
> > 
> > 
> > do turrets get the flash?
> 
> AI, nukie, and other turrets do not have a muzzle flash, sadly. That might be a bit more complex to add than standard guns.

ok yea i would think so, especially because you can put custom guns inside some turrets (which is super buggy on its own)

one more thing: dragnets? would they get muzzle flashes? their muzzles are massive, and they release bolas (pretty much) in one mode, and in the other some kind of teleportation thing

**Comment by WinterDarkraven** [comment]
> > > > > Make sure this doesn't affect [#78128 (comment)](https://github.com/tgstation/tgstation/pull/78128#issue-1879268689)
> > > > 
> > > > 
> > > > Thanks for the heads up. Will be removing muzzle flash from it!
> > > 
> > > 
> > > do turrets get the flash?
> > 
> > 
> > AI, nukie, and other turrets do not have a muzzle flash, sadly. That might be a bit more complex to add than standard guns.
> 
> ok yea i would think so, especially because you can put custom guns inside some turrets (which is super buggy on its own)
> 
> one more thing: dragnets? would they get muzzle flashes? their muzzles are massive, and they release bolas (pretty much) in one mode, and in the other some kind of teleportation thing

Dragsnet *do* have a muzzle flash! I actually never thought of them, but they use the default energy gun flash, which is cyan, which fits their colour pretty well! 

Xray and other guns, like pulse guns, have their own unique colour too. Green for xray and a darker blue for pulse. If you have any other guns I might have missed, throw 'em out there!

**Review by MrMelbert — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/gun.dm [hunk 1]
```diff
@@ -177,10 +184,22 @@
 	else
 		playsound(src, fire_sound, fire_sound_volume, vary_fire_sound)
 
+/obj/item/gun/proc/muzzle_flash_on()
+	if (can_muzzle_flash == TRUE)
+		set_light_on(TRUE)
+		addtimer(CALLBACK(src, PROC_REF(muzzle_flash_off)), light_time)
```
[the remark this answers is not in the document]
Comment (WinterDarkraven) [reply code/modules/projectiles/gun.dm:190]: There's a video showing a saw and rapid fire egun being shot, and it seems fine. If there's a better way that removes the timer, I'm happy to do that.

File: code/modules/projectiles/guns/energy.dm [hunk 2]
```diff
@@ -205,6 +206,16 @@
...
+	if (shot.select_name == "kill")
+	set_light_color(COLOR_SOFT_RED)
+	if (shot.select_name == "disable")
+	set_light_color(LIGHT_COLOR_CYAN)
+	if (shot.select_name == "DESTROY")
+	set_light_color(LIGHT_COLOR_BLUE)
+	if (shot.select_name == "stun")
+	set_light_color(LIGHT_COLOR_DIM_YELLOW)
+	if (shot.select_name == "ion")
+	set_light_color(LIGHT_COLOR_BLUE)
```
[the remark this answers is not in the document]
Comment (WinterDarkraven) [reply code/modules/projectiles/guns/energy.dm:218]: Working on this right now

**Comment by Admiral-Gigaton** [comment]
Tazer guns, like the hybrid taser, the pulse guns, the prototype gun at charliestation 

Uhhh... admin guns

The ninja modsuit and chrono modsuit both have module that shoot out these net things. Could you and/or would you try anything for those?

Tactical energy gun the ERTs sometimes get

The long ranged weaponry Wiki page

File: code/modules/projectiles/guns/ballistic.dm [hunk 3]
```diff
@@ -454,6 +455,7 @@
 	if(isitem(suppressed))
 		var/obj/item/I = suppressed
 		w_class -= I.w_class
+		can_muzzle_flash = TRUE //adds muzzle flash when suppressor is removed
```
[the remark this answers is not in the document]
Comment (carlarctg) [reply code/modules/projectiles/guns/ballistic.dm:458]:
```suggestion
		can_muzzle_flash = initial(can_muzzle_flash)
```

**Comment by Admiral-Gigaton** [comment]
Cannons, make sure the syndie and ghetto crossbows don't do it

Syndie ebow? The illegal tech variant?

Pka?

File: code/modules/projectiles/gun.dm [hunk 4]
```diff
@@ -21,7 +21,14 @@
 	item_flags = NEEDS_PERMIT
 	attack_verb_continuous = list("strikes", "hits", "bashes")
 	attack_verb_simple = list("strike", "hit", "bash")
-
+	light_range = 3
+	light_system = OVERLAY_LIGHT
```
Comment (MrMelbert) [inline code/modules/projectiles/gun.dm:25]: If you make this
```suggestion
	light_system = OVERLAY_LIGHT_DIRECTIONAL
```
Then the gun muzzle will be cone shaped in the direction of the gun shot, which might look nicer.

File: code/modules/projectiles/gun.dm (hunk 1 above)
[the remark this answers is not in the document]
Comment (MrMelbert) [reply code/modules/projectiles/gun.dm:190]: @LemonInTheDark Thoughts on this? I think it's *probably* fine but second opinion

File: code/modules/projectiles/gun.dm [hunk 5]
```diff
@@ -21,7 +21,14 @@
...
+	light_on = FALSE
+	light_color = LIGHT_COLOR_ORANGE
+	light_power = 0.5
```
Comment (LemonInTheDark) [inline code/modules/projectiles/gun.dm:28]: color will not show up very strongly with this power. You mentioned lowering it, why was that?

File: code/modules/projectiles/gun.dm [hunk 6]
```diff
@@ -177,10 +184,22 @@
...
+	if (can_muzzle_flash == TRUE)
+		set_light_on(TRUE)
+		addtimer(CALLBACK(src, PROC_REF(muzzle_flash_off)), light_time)
+	else
+		muzzle_flash_off()
+
+/obj/item/gun/proc/muzzle_flash_off()
+	set_light_on(FALSE)
+
+
```
Comment (LemonInTheDark) [inline code/modules/projectiles/gun.dm:197]: drop the extra newline

File: code/modules/projectiles/gun.dm (hunk 1 above)
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/modules/projectiles/gun.dm:190]: Performance wise I don't think it'd be a massive issue (we churn them a lot as is and this isn't a TON. I do kinda think we should give this TIMER_UNIQUE and TIMER_OVERRIDE tho. 

I guess it depends if you want firing 10 times a second to flash or just sustain on, but if I fire at 0 seconds, and fire again at 0.05 seconds, having it turn off after the 0.05 shot would likely feel weird (or worse if I fire again after 0.1, it'd turn off RIGHT after cause timers run after player input)

File: code/modules/projectiles/gun.dm (hunk 5 above)
Comment (WinterDarkraven) [reply code/modules/projectiles/gun.dm:28]: > color will not show up very strongly with this power. You mentioned lowering it, why was that?

The original version was a large, bright flash, which was horrible to look at. I'll play about with the numbers some more, directional lighting was suggested.

**Comment by SomeoneYouProbablyKnow** [comment]
This is brilliant. Love it

**Comment by jughu** [comment]
my epilepsy :(

**Comment by LemonInTheDark** [comment]
it's not THAT strong, but if that's a serious and not memey post we might be able to do something about it. It's annoying though cause it's tied to the lighting plane, hard to disable that on a player by player basis.

**Comment by projectkepler-ru** [comment]
Amazing
can we get buckshot and slug back for security too while at it :3

**Comment by WinterDarkraven** [comment]
I'm closing this PR due to, well, issues. Basically, no time to work on this right now! 

Don't really know how this works, so... probably make a new PR when I get around to this?

**Comment by LemonInTheDark** [comment]
It's good shit lookin forward to it being cleaned up if possible.
Also yeah you could reopen this but it's much less likely to be seen, so just a new pr is fine.

**Pull request closed without merging**